pkg/shell/terraform.go

```diff
@@ -136,9 +148,29 @@ func getOutputs(tf *tfexec.Terraform) (map[string]cty.Value, error) {
 		}
 	}
 
+	var apply bool
 	if wantsChange {
-		return map[string]cty.Value{},
-			fmt.Errorf("cloud infrastructure requires changes; please run \"terraform -chdir=%s apply\"", tf.WorkingDir())
+		log.Println("cloud infrastructure requires changes")
+		switch applyBehavior {
+		case AutomaticApply:
+			apply = true
+		case PromptBeforeApply:
+			apply = AskForConfirmation(fmt.Sprintf("Do you want to deploy group %s", tf.WorkingDir()))
```

Comment:
I feel there would be a lot of value to print the plan before asking if we should apply. If there is a way for us to add this, even if it is not efficient to start, I think we should as it totally changes the user experience.

My recommendation to not overburden this initial implementation would be to call plan again, now to output, under the circumstances that changes are needed and we need user approval. 

I think you could simply add the following before asking the user to approve:

```
		log.Printf("generating plan for group %s", tf.WorkingDir())
		tf.SetStdout(os.Stdout)
		tf.SetStderr(os.Stderr)
		tf.Plan(context.Background()) // guessing on syntax
		tf.SetStdout(nil)
		tf.SetStderr(nil)
```

User output would look like:

```
initializing terraform directory igc-01/one
testing if terraform state of igc-01/one is in sync with cloud infrastructure
cloud infrastructure requires changes
generating plan for group igc-01/one
...
... many lines of terraform plan
...
Do you want to deploy group igc-01/one [y/n]:
```

Then we can optimize so we are not running plan twice in future pr.

Reply:
I can do this, but will probably choose the "optimal" approach of directing the plan to a file and running apply on the file. Lots of good reasons and not obviously much more effort than running twice.

Reply:
I've chosen the following approach:

- capture the stdout of plan (in human-readable format)
- prompt the user with 4 options: `Display proposed changes, Apply proposed changes, Stop and exit, Continue without applying? [d,a,s,c]`
- Loop forever until user picks 1 of the 4 options.

Logic behind making display of the plan optional is that it is very long and causes me to become disoriented if printed immediately to screen.

There is a sub-optimality in that there does not appear to be an existing func in tfexec that will capture both human and machine readable format for the plan. So there is no guarantee that the apply stage will attempt _exactly_ the changes being proposed. I will address this with a task for future work.

Reply:
I disagree with requiring the user to take action before they are told what is about to change (in this case the `d` option). In addition we are now providing the user with 4 options here.

Terraform's native behavior is to always print the delta. I don't believe we should deviate from this behavior without good reason. 

I could see at some point in the future only printing the summary line (`x added, x modified, x deleted`) and having the user ask for the full output, but I don't think that should be our starting point.

Reply:
@cboneti could you help weight in here as a tiebreaker.

Reply:
both of you  are bringing important points here. 
I believe it would be important to give a sense of the changes, so this is what I propose, do one of the two options as a starting point. Pick what you prefer or whatever is easier: 
1) Print, as Nick pointed, a summary line as described below: 
```
❯❯❯ ghpc deploy hpc-small
2023/05/10 14:31:05 initializing terraform directory hpc-small/primary
2023/05/10 14:31:23 testing if terraform state of hpc-small/primary is in sync with cloud infrastructure
2023/05/10 14:31:30 cloud infrastructure requires changes:

Plan: 13 to add, 0 to change, 0 to destroy.

Display proposed changes, Apply proposed changes, Stop and exit, Continue without applying? [d,a,s,c]:
...
```

or, if it is easier this way: 
2) always print terraform plan output and then adjust the prompt accordingly (notice changes after tf plan): 
```
❯❯❯ ghpc deploy hpc-small
2023/05/10 14:31:05 initializing terraform directory hpc-small/primary
2023/05/10 14:31:23 testing if terraform state of hpc-small/primary is in sync with cloud infrastructure
##
## MANY LINES OF TERRAFORM PLAN OUTPUT
##
2023/05/10 14:31:30 cloud infrastructure requires changes for hpc-small/primary
Apply proposed changes, Stop and exit, Continue without applying? [a,s,c]: a
2023/05/10 14:31:35 running terraform apply on group hpc-small/primary
module.slurm_login.module.slurm_cluster_login_node.data.google_compute_default_service_account.default: Reading...
...
```

Reply:
Offline discussion resulted in choosing (1). 3c598f29 includes:

- printing a summary of changes immediately upon prompt
- continue to allow an optional full summary by entering `d`

Because it was simple within this effort of work, I added the functionality of applying the plan actually being shown to the user rather than "crossing your fingers" that nothing has changed by the time apply is run.

Edge cases addressed:

- Packer full and summary of changes are the same
- Future task will be created to use `PlanJSON` and parse the output for the [change_summary](https://developer.hashicorp.com/terraform/internals/machine-readable-ui#change-summary) field instead of regexp for `Plan:`. If no match is found, an advisory to print the full proposed changes is written to screen.

Reply:
LGTM, please assign to Nick

Reply:
LGTM. Just want to test then will approve.